File: app/src/androidTest/java/com/example/win/easy/view/lock/SearchFragmentTest.java
package com.example.win.easy.view.lock;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.fragment.app.FragmentFactory;
import androidx.fragment.app.testing.FragmentScenario;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;

import com.example.win.easy.R;
import com.example.win.easy.display.DisplayServiceAdapter;
import com.example.win.easy.tool.SongListWithSongs;
import com.example.win.easy.value_object.SongListVO;
import com.example.win.easy.value_object.SongVO;
import com.example.win.easy.viewmodel.SongListViewModel;
import com.example.win.easy.viewmodel.SongViewModel;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.example.win.easy.enumeration.DataSource.Local;
import static com.example.win.easy.enumeration.DataSource.WangYiYun;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4ClassRunner.class)
public class SearchFragmentTest {

    @Test
    public void testNormalSelection(){
        search(testSequence);
        search(testSequence);
        verifySongListDisplayed(localMatchedSongs,neteaseMatchedSongs);

        clickTab(neteaseMatchedSongs);
        clickSong(songInHateAndNormal);
        verifySongListDisplayed(hate,normal);
        verifyMediaIsPlaying(songInHateAndNormal,hate);

        clickTab(normal);
        verifyMediaIsPlaying(songInHateAndNormal,normal);

        clickSong(songInNormalAndLove);
        verifySongListDisplayed(normal,love);
        verifyMediaIsPlaying(songInNormalAndLove,normal);

        clickTab(love);
        verifyMediaIsPlaying(songInNormalAndLove,love);
    }

    @Test
    public void testSequenceNoSongsMatch(){
        search(sequenceNoSongsMatch);
    }

    @Test
    public void testSelectSongNotInAnySongList(){
        search(testSequence);
        verifySongListDisplayed(localMatchedSongs,neteaseMatchedSongs);

        clickSong(songNotInAnySongList);
        verifySongListDisplayed(songListOfAllSong);
        verifyMediaIsPlaying(songNotInAnySongList,songListOfAllSong);
    }

    private void verifyMediaIsPlaying(SongVO beingPlayedSong, SongListWithSongs beingPlayedSongList) {
        assertEquals(beingPlayedSong,displayServiceAdapter.getCurrentDisplayedSong());
        assertEquals(beingPlayedSongList.getSongs(),displayServiceAdapter.getCurrentDisplayedSongList());
    }

    private void search(List<Character> sequence){
        scenario.onFragment(fragment -> fragment.search(sequence));
    }

    private void verifySongListDisplayed(SongListWithSongs... filledSongLists){
        SongListWithSongs theFirstOne=filledSongLists[0];
        int length=filledSongLists.length;
        for (int songListIndex=0;songListIndex<length;songListIndex++){
            SongListWithSongs songListWithSongs=filledSongLists[songListIndex];
            onView(withText(songListWithSongs.getSongList().getName())).check(matches(isDisplayed()));
            clickTab(songListWithSongs);


//            List<SongVO> songs=songListWithSongs.getSongs();
//            int songListSize=songs.size();
//            for (int songIndex=0;songIndex<songListSize;songIndex++) {
//                SongVO songVO=songs.get(songIndex);
//                onView(allOf(withText(songVO.getName()),isDescendantOfA(nthChildOf(nthChildOf(hasSibling(withChild(withChild(withChild(withText(songListWithSongs.getSongList().getName()))))),songListIndex),songIndex+1)))).check(matches(isDisplayed()));
//            }
        }
        clickTab(theFirstOne);
    }

    private void clickTab(SongListWithSongs songListWithSongs){
        onView(withText(songListWithSongs.getSongList().getName())).perform(click());
    }

    private void clickSong(SongVO song){
        onView(withText(song.getName())).perform(click());
    }

    @Before
    public void setUp() throws FailToReplaceSongListException {
        MockitoAnnotations.initMocks(this);

        setUpData();

        mockDisplayService();
        mockViewModel();
        mockViewModelFactory();
        mockFragmentFactory();

        launchTestFragment();
    }

    private void setUpData(){
        songListOfAllSong.getSongs().add(songNotInAnySongList);
        songListOfAllSong.getSongs().add(songOnlyInHate);
        songListOfAllSong.getSongs().add(songInHateAndNormal);
        songListOfAllSong.getSongs().add(songInNormalAndLove);

        localMatchedSongs.getSongs().add(songNotInAnySongList);
        localMatchedSongs.getSongs().add(songOnlyInHate);

        neteaseMatchedSongs.getSongs().add(songInHateAndNormal);
        neteaseMatchedSongs.getSongs().add(songInNormalAndLove);

        hate.getSongs().add(songOnlyInHate);
        hate.getSongs().add(songInHateAndNormal);

        normal.getSongs().add(songInHateAndNormal);
        normal.getSongs().add(songInNormalAndLove);

        love.getSongs().add(songInNormalAndLove);

        allSongsInRepoLiveData.postValue(songListOfAllSong.getSongs());
    }

    private void mockDisplayService() throws FailToReplaceSongListException {
        doAnswer(invocation -> {
            SongVO songArg=invocation.getArgument(0);
            List<SongVO> songListArg=invocation.getArgument(1);

            doReturn(songArg).when(displayServiceAdapter).getCurrentDisplayedSong();
            doReturn(songListArg).when(displayServiceAdapter).getCurrentDisplayedSongList();
            return null;
        }).when(displayServiceAdapter).startWith(any(SongVO.class),any(List.class));
        doAnswer(invocation -> {
            List<SongVO> songListArg=invocation.getArgument(0);

            doReturn(songListArg).when(displayServiceAdapter).getCurrentDisplayedSongList();
            return null;
        }).when(displayServiceAdapter).replaceSongList(any(List.class));
    }

    private void mockViewModel(){
        when(songViewModel.songsMatch(testSequence)).thenReturn(songListOfAllSong.getSongs());
        when(songViewModel.songsMatch(sequenceNoSongsMatch)).thenReturn(new ArrayList<>());

        when(songViewModel.songListsContain(songNotInAnySongList)).thenReturn(new ArrayList<>());
        when(songViewModel.songListsContain(songOnlyInHate)).thenReturn(Collections.singletonList(hate.getSongList()));
        when(songViewModel.songListsContain(songInHateAndNormal)).thenReturn(Arrays.asList(hate.getSongList(),normal.getSongList()));
        when(songViewModel.songListsContain(songInNormalAndLove)).thenReturn(Arrays.asList(normal.getSongList(),love.getSongList()));

        doReturn(allSongsInRepoLiveData).when(songViewModel).getAllSongs();

        when(songListViewModel.songsOf(hate.getSongList())).thenReturn(hate.getSongs());
        when(songListViewModel.songsOf(normal.getSongList())).thenReturn(normal.getSongs());
        when(songListViewModel.songsOf(love.getSongList())).thenReturn(love.getSongs());
    }

    private void mockViewModelFactory(){
        when(viewModelFactory.create(SongViewModel.class)).thenReturn(songViewModel);
        when(viewModelFactory.create(SongListViewModel.class)).thenReturn(songListViewModel);
    }

    private void mockFragmentFactory(){
        when(fragmentFactory.instantiate(any(ClassLoader.class),eq(SearchFragment.class.getName()))).thenReturn(searchFragment);
    }

    private void launchTestFragment(){
        scenario=FragmentScenario.launchInContainer(SearchFragment.class,null, R.style.AppTheme,fragmentFactory);
    }

    @Rule public InstantTaskExecutorRule instantTaskExecutorRule=new InstantTaskExecutorRule();
    @InjectMocks SearchFragment searchFragment;
    @Mock FragmentFactory fragmentFactory;
    @Mock SongListViewModel songListViewModel;
    @Mock SongViewModel songViewModel;
    @Mock ViewModelProvider.Factory viewModelFactory;
    @Mock DisplayServiceAdapter displayServiceAdapter;

    private FragmentScenario<SearchFragment> scenario;

    private SongListWithSongs songListOfAllSong=SongListWithSongs.builder().songList(SongListVO.builder().name("所有歌曲").build()).songs(new ArrayList<>()).build();
    private SongListWithSongs localMatchedSongs=SongListWithSongs.builder().songList(SongListVO.builder().name(Local.name()).dataSource(Local).build()).songs(new ArrayList<>()).build();
    private SongListWithSongs neteaseMatchedSongs=SongListWithSongs.builder().songList(SongListVO.builder().name(WangYiYun.name()).dataSource(WangYiYun).build()).songs(new ArrayList<>()).build();
    private SongListWithSongs hate=SongListWithSongs.builder().songList(SongListVO.builder().name("讨厌").dataSource(Local).build()).songs(new ArrayList<>()).build();
    private SongListWithSongs normal=SongListWithSongs.builder().songList(SongListVO.builder().name("一般般").dataSource(WangYiYun).build()).songs(new ArrayList<>()).build();
    private SongListWithSongs love=SongListWithSongs.builder().songList(SongListVO.builder().name("喜欢！").dataSource(WangYiYun).build()).songs(new ArrayList<>()).build();

    private SongVO songNotInAnySongList=SongVO.builder().name("无法地带").source(Local).build();
    private SongVO songOnlyInHate=SongVO.builder().name("只在讨厌里").source(Local).build();
    private SongVO songInHateAndNormal =SongVO.builder().name("在讨厌和一般般里").source(WangYiYun).build();
    private SongVO songInNormalAndLove=SongVO.builder().name("在一般般和喜欢里").source(WangYiYun).build();

    private MutableLiveData<List<SongVO>> allSongsInRepoLiveData=new MutableLiveData<>();
    private List<Character> sequenceNoSongsMatch= Arrays.asList('A','B','C');
    private List<Character> testSequence= Arrays.asList('E','F','G');
}

File: app/src/androidTest/java/com/example/win/easy/view/lock/SearchFragmentTest.kt
package com.example.win.easy.view.lock

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.fragment.app.FragmentFactory
import androidx.fragment.app.testing.FragmentScenario
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import com.example.win.easy.R
import com.example.win.easy.TestCoroutineRule
import com.example.win.easy.display.DisplayService
import com.example.win.easy.enumeration.DataSource
import com.example.win.easy.db.SongDO
import com.example.win.easy.db.SongListDO
import com.example.win.easy.tool.SongListWithSongs
import com.example.win.easy.viewmodel.SongListViewModel
import com.example.win.easy.viewmodel.SongViewModel
import io.mockk.CapturingSlot
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.runBlocking
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import kotlin.collections.ArrayList

@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4ClassRunner::class)
class SearchFragmentTest {
    @Test
    fun testNormalSelection()= runBlocking {
        search(testSequence)
        verifySongListDisplayed(mLocalMatchedSongsC, mOnlineMatchedSongsC)
        clickTab(mOnlineMatchedSongsC)
        clickSong(songInHateAndNormal)
        verifySongListDisplayed(mNormalC,mHateC)
        clickTab(mHateC)
        verifyMediaIsPlaying(songInHateAndNormal, mHateC)
        clickTab(mNormalC)
        verifyMediaIsPlaying(songInHateAndNormal, mNormalC)
        clickSong(songInNormalAndLove)
        verifySongListDisplayed(mNormalC, mLoveC)
        verifyMediaIsPlaying(songInNormalAndLove, mNormalC)
        clickTab(mLoveC)
        verifyMediaIsPlaying(songInNormalAndLove, mLoveC)
    }

    @Test
    fun testSequenceNoSongsMatch() {
        search(sequenceNoSongsMatch)
    }

    @Test
    fun testSelectSongNotInAnySongList() {
        search(testSequence)
        verifySongListDisplayed(mLocalMatchedSongsC, mOnlineMatchedSongsC)
        clickSong(songNotInAnySongList)
        verifySongListDisplayed(mAllSongsC)
    }

    private fun verifyMediaIsPlaying(beingPlayedSong: SongDO, beingPlayedSongList: SongListWithSongs) {
        assertEquals(beingPlayedSong, displayService.currentSong())
        assertEquals(beingPlayedSongList.songList, displayService.currentSongList())
        assertEquals(beingPlayedSongList.songs, displayService.currentSongs())
    }

    private fun search(sequence: List<Char>) {
        scenario.onFragment { fragment: SearchFragment -> fragment.search(sequence) }
    }

    private fun verifySongListDisplayed(vararg filledSongLists: SongListWithSongs) {
        for (songListWithSongs in filledSongLists) {
            Espresso.onView(ViewMatchers.withText(songListWithSongs.name)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
            clickTab(songListWithSongs)
        }
        clickTab(filledSongLists[0])
    }

    private fun clickTab(songListWithSongs: SongListWithSongs) {
        Espresso.onView(ViewMatchers.withText(songListWithSongs.name)).perform(ViewActions.click())
    }

    private fun clickSong(song: SongDO) {
        Espresso.onView(ViewMatchers.withText(song.name)).perform(ViewActions.click())
    }

    @Before
    @Throws(FailToReplaceSongListException::class)
    fun setUp() {
        MockKAnnotations.init(this)
        mockDisplayService()
        mockViewModelFactory()
        mockViewModel()
        mockFragmentFactory()
        launchTestFragment()
    }

    private val songCap=CapturingSlot<SongDO>()
    private val songListCap=CapturingSlot<SongListDO>()
    private val songsCap=CapturingSlot<List<SongDO>>()
    private val replaceListCap=CapturingSlot<SongListDO>()
    private val replaceSongsCap=CapturingSlot<List<SongDO>>()

    @Throws(FailToReplaceSongListException::class)
    private fun mockDisplayService() {
        every { displayService.restartWith(capture(songCap),capture(songListCap),capture(songsCap)) }answers {
            every { displayService.currentSong() } returns songCap.captured
            every { displayService.currentSongList() } returns songListCap.captured
            every { displayService.currentSongs() } returns songsCap.captured
        }
        every { displayService.configDisplayList(capture(replaceListCap),capture(replaceSongsCap)) } answers {
            every { displayService.currentSongList() } returns replaceListCap.captured
            every { displayService.currentSongs() } returns replaceSongsCap.captured
        }
    }

    private val songDOCap=CapturingSlot<(List<SongDO>)->Unit>()
    private val hateCap=CapturingSlot<(List<SongListDO>)->Unit>()
    private val hateAndNormalCap=CapturingSlot<(List<SongListDO>)->Unit>()
    private val normalAndLoveCap=CapturingSlot<(List<SongListDO>)->Unit>()
    private val listOfSongListCap=CapturingSlot<(List<SongListDO>)>()
    private val songsInListsBlockCap=CapturingSlot<(List<List<SongDO>>)->Unit>()

    private fun mockViewModel() {
        every { songViewModel.launchLoadSongsBySeq(testSequence,capture(songDOCap)) } answers{songDOCap.captured(mAllSongsC.songs)}
        every { songViewModel.launchLoadSongsBySeq(sequenceNoSongsMatch,captureLambda()) } answers{ lambda<(List<SongDO>)->Unit>().captured(ArrayList())}

        every { songListViewModel.launchLoadSongListsBySong(songNotInAnySongList,captureLambda()) } answers{lambda<(List<SongDO>)->Unit>().captured(ArrayList())}
        every { songListViewModel.launchLoadSongListsBySong(songOnlyInHate,capture(hateCap)) } answers{hateCap.captured(listOf(mHateC.songList))}
        every { songListViewModel.launchLoadSongListsBySong(songInHateAndNormal,capture(hateAndNormalCap)) } answers{hateAndNormalCap.captured(listOf(mHateC.songList, mNormalC.songList))}
        every { songListViewModel.launchLoadSongListsBySong(songInNormalAndLove,capture(normalAndLoveCap)) } answers{normalAndLoveCap.captured(listOf(mNormalC.songList, mLoveC.songList))}

        every { songViewModel.allSongs } returns allSongsInRepoLiveData

        every { songViewModel.launchLoadSongsBySongLists(capture(listOfSongListCap),capture(songsInListsBlockCap)) }.answers{
            mutableListOf<List<SongDO>>().run {
                for (songList in listOfSongListCap.captured)
                    add(when(songList){
                        mHateC.songList-> mHateC.songs
                        mNormalC.songList-> mNormalC.songs
                        mLoveC.songList-> mLoveC.songs
                        mAllSongsC.songList-> mAllSongsC.songs
                        mLocalMatchedSongsC.songList-> mLocalMatchedSongsC.songs
                        mOnlineMatchedSongsC.songList-> mOnlineMatchedSongsC.songs
                        else -> listOf()
                    })
                songsInListsBlockCap.captured(this)
            }
        }
    }

    private fun mockViewModelFactory() {
        every { viewModelFactory.create(SongViewModel::class.java)} returns (songViewModel)
        every { viewModelFactory.create(SongListViewModel::class.java)} returns (songListViewModel)
    }

    private fun mockFragmentFactory() {
        every { fragmentFactory.instantiate(any(), any())} returns (searchFragment)
    }

    private fun launchTestFragment() {
        scenario = FragmentScenario.launchInContainer(SearchFragment::class.java, null, R.style.AppTheme, fragmentFactory)
    }

    @get:Rule val instantTaskExecutorRule = InstantTaskExecutorRule()
    @get:Rule val testCoroutineRule = TestCoroutineRule()

    @InjectMockKs lateinit var searchFragment: SearchFragment
    @MockK lateinit var fragmentFactory: FragmentFactory
    @RelaxedMockK lateinit var songListViewModel: SongListViewModel
    @RelaxedMockK lateinit var songViewModel: SongViewModel
    @MockK lateinit var viewModelFactory: ViewModelProvider.Factory
    @MockK lateinit var displayService: DisplayService

    private lateinit var scenario: FragmentScenario<SearchFragment>

    private val songNotInAnySongList = SongDO(name = "无法地带", source = DataSource.Local)
    private val songOnlyInHate = SongDO(name = "只在讨厌里", source = DataSource.Local)
    private val songInHateAndNormal = SongDO(name = "在讨厌和一般般里", source = DataSource.WangYiYun,uid=1,remoteId = 3)
    private val songInNormalAndLove = SongDO(name = "在一般般和喜欢里", source = DataSource.WangYiYun,uid=1,remoteId = 4)

    private val mAllSongsC = SongListWithSongs(songList= SongListDO(name = "所有歌曲", source = DataSource.Local),songs= listOf(songNotInAnySongList,songOnlyInHate,songInHateAndNormal,songInNormalAndLove))
    private val mLocalMatchedSongsC = SongListWithSongs(songList= SongListDO(name = DataSource.Local.name, source = DataSource.Local),songs= listOf(songNotInAnySongList,songOnlyInHate))
    private val mOnlineMatchedSongsC = SongListWithSongs(songList= SongListDO(name = DataSource.WangYiYun.name, source = DataSource.WangYiYun),songs= listOf(songInHateAndNormal,songInNormalAndLove))
    private val mHateC = SongListWithSongs(songList= SongListDO(name = "讨厌", source = DataSource.Local),songs= listOf(songOnlyInHate,songInHateAndNormal))
    private val mNormalC = SongListWithSongs(songList= SongListDO(name = "一般般", source = DataSource.WangYiYun,uid=1,remoteId = 1),songs= listOf(songInHateAndNormal,songInNormalAndLove))
    private val mLoveC = SongListWithSongs(songList= SongListDO(name = "喜欢！", source = DataSource.WangYiYun,uid=1,remoteId = 2),songs= listOf(songInNormalAndLove))


    private val allSongsInRepoLiveData = MutableLiveData<List<SongDO>>().apply { postValue(mAllSongsC.songs) }

    private val sequenceNoSongsMatch = listOf('A', 'B', 'C')
    private val testSequence = listOf('E', 'F', 'G')
}